src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -51,200 +64,177 @@ public Task WarmUpCacheAsync(Project? project, CancellationToken cancellationTok
             return (getCacheResults.SelectAsArray(GetItemsFromCacheResult), isPartialResult);
 
-            ImmutableArray<CompletionItem> GetItemsFromCacheResult(GetCacheResult cacheResult)
-            {
-                return cacheResult.Entry.GetItemsForContext(
-                         Language,
-                         GenericTypeSuffix,
-                         currentCompilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(cacheResult.Assembly),
-                         syntaxContext.IsAttributeNameContext,
-                         IsCaseSensitive,
-                         options.HideAdvancedMembers);
-            }
+            ImmutableArray<CompletionItem> GetItemsFromCacheResult(CacheEntry cacheEntry)
+                => cacheEntry.GetItemsForContext(
+                    currentCompilation,
+                    Language,
+                    GenericTypeSuffix,
+                    syntaxContext.IsAttributeNameContext,
+                    IsCaseSensitive,
+                    options.HideAdvancedMembers);
         }
 
-        private async Task<(ImmutableArray<GetCacheResult> results, bool isPartial)> GetCacheEntriesAsync(Project currentProject, bool forceCacheCreation, CancellationToken cancellationToken)
+        private async Task<(ImmutableArray<CacheEntry> results, bool isPartial)> GetCacheEntriesAsync(Project currentProject, bool forceCacheCreation, CancellationToken cancellationToken)
         {
             var isPartialResult = false;
-            var _ = ArrayBuilder<GetCacheResult>.GetInstance(out var builder);
-
-            var currentCompilation = await currentProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var editorBrowsableInfo = new Lazy<EditorBrowsableInfo>(() => new EditorBrowsableInfo(currentCompilation));
-
-            var cacheResult = await GetCacheForProjectAsync(currentProject, forceCacheCreation: true, editorBrowsableInfo, cancellationToken).ConfigureAwait(false);
-
-            // We always force create a cache for current project.
-            Contract.ThrowIfFalse(cacheResult.HasValue);
-            builder.Add(cacheResult.Value);
+            using var _1 = ArrayBuilder<CacheEntry>.GetInstance(out var resultBuilder);
+            using var _2 = ArrayBuilder<Project>.GetInstance(out var projectsBuilder);
+            using var _3 = PooledHashSet<ProjectId>.GetInstance(out var nonGlobalAliasedProjectReferencesSet);
 
             var solution = currentProject.Solution;
             var graph = solution.GetProjectDependencyGraph();
-            var referencedProjects = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id).SelectAsArray(id => solution.GetRequiredProject(id));
+            var referencedProjects = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id).Select(id => solution.GetRequiredProject(id)).Where(p => p.SupportsCompilation);
+
+            projectsBuilder.Add(currentProject);
```

Comment:
Note that we no longer have any special treatment for current project, i.e. we return whatever in the cache, or if the cache isn't ready, we don't return anything from it.